### PR TITLE
SDI-266 TAP return refactor and return to different prison

### PR DESCRIPTION
**Pull request by andymarke**
1. Existing TAP return moved to new service to share code
2. Existing TAP return bugs fixed by moving to new shared services
3. New functionality, return from TAP to different prison based on court return functionality

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt [hunk 1]
```diff
@@ -144,6 +145,86 @@ class ExternalMovementTransferService(
       )
     }
   }
+  fun updateMovementsForTransferInAfterTemporaryAbsenceToSamePrison(
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferService.kt:148]: TODO plan to refactor this class, basically most of the methods do roughly the same thing

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffendersResourceTransferImpTest.kt [hunk 2]
```diff
@@ -1398,14 +1397,14 @@ class OffendersResourceTransferImpTest : ResourceTest() {
       @DisplayName("When bed is released")
       inner class BedReleased {
         private lateinit var transferOutDateTime: LocalDateTime
-        private val toAddressId = 18248
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffendersResourceTransferImpTest.kt:1401]: test was quite right should have done a TAP release to a city

File: src/test/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferServiceTest.kt [hunk 3]
```diff
@@ -441,4 +455,448 @@ internal class ExternalMovementTransferServiceTest {
       }
     }
   }
+
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/ExternalMovementTransferServiceTest.kt:458]: used same pattern for the other tests in the class - not very interesting and not giving a lot of value

File: src/test/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonTransferServiceTest.kt [hunk 4]
```diff
@@ -64,6 +69,12 @@ internal class PrisonTransferServiceTest {
   private val bookingLastMovementCourtWithEventId =
     getMovement(toAgencyIn = toCourt, movementReasonCode = "CRT", movementTypeCode = "CRT", eventIdIn = 123)
 
+  private val bookingLastMovementTAP =
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonTransferServiceTest.kt:72]: used same pattern for the other tests in the class - not very interesting and not giving a lot of value given the integration test covers the same scenarios

File: src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonTransferService.kt [hunk 5]
```diff
@@ -152,8 +156,82 @@ class PrisonTransferService(
     return transformer.transform(booking)
   }
 
-  fun transferViaTemporaryAbsence(): InmateDetail {
-    TODO()
+  fun transferInAfterTemporaryAbsence(offenderNo: String, request: RequestForTemporaryAbsenceArrival): InmateDetail {
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonTransferService.kt:159]: meat of the change HERE

**Review by lbennett-moj — APPROVED** [review]
LGTM

**Pull request merged**